### ISSUE TITLE
train/val/test splits

### ISSUE THREAD

**Issue by jsadler2** [USGS-R/drb-do-ml#9]
We will need to come up with splits for train/validation/test splits.

**Comment by galengorski** [comment]
Did stream temperature use a cross validation approach for training? We'll also need to think about this for hyperparameter tuning

**Comment by jsadler2** [comment]
they held out some "validation" data. I don't know if they did any "cross-validation". 

And yes. we will definitely need to think about this for hyperparameter tuning.

**Comment by jsadler2** [comment]
The temp project did quite a bit of work when they decided on their split: https://github.com/USGS-R/delaware-model-prep/issues/93

**Comment by jsadler2** [comment]
I made some figures that show some overall data availability over time and by site:

### Daily sites (daily mean DO)
![daily_daily_means](https://user-images.githubusercontent.com/6943441/144119461-0d1300d6-1f5b-4c5b-8ad5-c15b1900819e.jpg)

### Instantaneous sites (daily mean DO)
![inst_daily_means](https://user-images.githubusercontent.com/6943441/144119578-76bd19e9-8e72-4e46-9ebb-210c6c7f29fd.jpg)

I also made a figure of the DOY averages across the sites:
![doy_means](https://user-images.githubusercontent.com/6943441/144119626-509097df-3617-4710-bf41-c89a875d4875.jpg)

**Comment by jsadler2** [comment]
And here's a map of the points with the most well-observed points labeled:
![image](https://user-images.githubusercontent.com/6943441/144123680-f02c044a-98c6-4ede-82fe-321e642179b7.png)

**Comment by jsadler2** [comment]
There's basically 5 sites with consistent "daily" data: 
![image](https://user-images.githubusercontent.com/6943441/144128360-39796db8-9607-48f8-bce7-73232bfb753d.png)
![image](https://user-images.githubusercontent.com/6943441/144128493-b95a96b0-dbeb-4502-b686-5682c91ee102.png)

**Comment by jsadler2** [comment]
###  filtering to just the most well-observed instantaneous sites
And the 12 sites that have 300+ (arbitrary threshold) day-observations from the instantaneous record:
![image](https://user-images.githubusercontent.com/6943441/144129585-5a71cdec-5f35-49e1-99bd-b4408368465c.png)
![image](https://user-images.githubusercontent.com/6943441/144129643-369449c1-66c3-4284-8213-395c9e062bd7.png)

**Comment by lekoenig** [comment]
It's helpful to see these time series plotted together. What do you think about the daily and inst sites being separated so far in time (e.g. 1980's and then 2008-present)? Does this affect the value of the daily-site data, or how we might decide on the train/test splits?

**Comment by jsadler2** [comment]
I don't think it affects their value because they don't overlap. I was thinking that we could just use the daily sites in the training dataset and not the validation or testing. But there are some differences in the plots from the updated data that I will post below that changes the story a little bit.

**Comment by jsadler2** [comment]
### Updated daily plot
After merging in #23, I reran the code to generate the plots above and got a different figure from the one above for the daily data:
![image](https://user-images.githubusercontent.com/6943441/145096992-c1b0838a-32b4-4dab-9aa8-75dd1d2296a0.png)

**Comment by jsadler2** [comment]
There are several more stations in the one I just posted. (18 before, now 29). I think that the additional ones that show up here are sites that before only showed up in the instantaneous plots (e.g., 01477050). So that makes sense to me. One thing I'm curious about, though, is why is there now data at those sites from earlier years when in the instantaneous data record, we only see the data starting at ~2008.

**Comment by lekoenig** [comment]
You're right - previously, I was prioritizing sites that had instantaneous data and so was not including data from sites with both "dv" and "uv" values (those sites would have only showed up in the instantaneous data). However, because the time frames of data collection can vary by quite a bit I've since opted to download the "dv" data, too. 

I don't have a definitive answer to your second question other than to say that instantaneous values collected prior to October 1, 2007 are not publicly available on NWIS. My understanding is that this has to do with the reliability of oxygen sensors during continuous deployments using older technology, however, I'm not certain that that's the reason for the 2007 cutoff.

**Comment by jsadler2** [comment]
> because the time frames of data collection can vary by quite a bit I've since opted to download the "dv" data, too.

Got it. 👍 

>that instantaneous values collected prior to October 1, 2007 are not publicly available on NWIS. My understanding is that this has to do with the reliability of oxygen sensors during continuous deployments using older technology, however, I'm not certain that that's the reason for the 2007

Ah. Interesting. That explanation makes sense to me.

**Comment by jsadler2** [comment]
Okay. Here's what I'm proposing for splits.

Note I did not consider the sparsely monitored sites because I do not think they will be as impactful on the model performance as the sites that are well-monitored (I took "well-monitored" to mean 300 points).

**Time periods:** 
- Validation: 2017-01-01 - 2018-12-31
- Testing: 2019-01-01 - present

**Sites:**
- Validation: `01475530`, `01475538` 
- Testing: `01475530`, `01475538`

Assuming we can get met data through 2020 (which I think will be doable), that makes the following portions of data for each partition:

train|val|test
-|-|-
 0.76 | 0.17 | 0.07

Here are some plots about how this would look:
![image](https://user-images.githubusercontent.com/6943441/145652287-d877f605-bdca-49ae-978f-8af43317f6be.png)
![image](https://user-images.githubusercontent.com/6943441/145652292-09d1eda1-9b89-4ac2-9e78-f5ea00422897.png)

![image](https://user-images.githubusercontent.com/6943441/145652330-adef8a76-f22f-43ff-a165-8387a3fa05a8.png)
There's a blue dot `01472104` hiding just behind the green dot there. Those two sites are very close but do not have overlapping time periods. ... and now that I'm writing this, I'm wondering if that will be that helpful...
